Add fmi4cpp

traversaro opened the pull request:
This PR adds a package for fmi4cpp (https://github.com/NTNU-IHB/FMI4cpp), a C++ implementation of the Functional Mock-up Interface (https://fmi-standard.org/) standard.

Other libraries that implement this standard already available in conda-forge:
* https://github.com/conda-forge/fmilib-feedstock
* https://github.com/conda-forge/fmpy-feedstock

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/fmi4cpp```) and found it was in an excellent condition.

traversaro commented:
macOS is failing with:
~~~
$SRC_DIR/src/fmi4cpp/fmi2/fmu.cpp:42:37: error: 'path' is unavailable: introduced in macOS 10.15
    const std::string fmuName = fs::path(fmuPath).stem().string();
                                    ^
$BUILD_PREFIX/bin/../include/c++/v1/filesystem:909:24: note: 'path' has been explicitly marked unavailable here
                                ^
$BUILD_PREFIX/bin/../include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
$SRC_DIR/src/fmi4cpp/fmi2/fmu.cpp:42:51: error: 'stem' is unavailable: introduced in macOS 10.15
    const std::string fmuName = fs::path(fmuPath).stem().string();
                                                  ^
$BUILD_PREFIX/bin/../include/c++/v1/filesystem:1368:34: note: 'stem' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY path stem() const { return string_type(__stem()); }
                                 ^
$SRC_DIR/src/fmi4cpp/fmi2/fmu.cpp:42:51: error: '~path' is unavailable: introduced in macOS 10.15
    const std::string fmuName = fs::path(fmuPath).stem().string();
                                                  ^
$BUILD_PREFIX/bin/../include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
~~~

Disabling it and re-enabling it as soon as the recipe is merged.

traversaro commented:
CI is happy and the PR is ready for review. https://github.com/orgs/conda-forge/teams/help-c-cpp

traversaro commented:
> CI is happy and the PR is ready for review. https://github.com/orgs/conda-forge/teams/help-c-cpp

Not sure why, but the reference was not inserted correctly, the recipe is ready for review @conda-forge/help-c-cpp .

carterbox commented on recipes/fmi4cpp/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - libzip
+    - boost-cpp
+  run:
+    - boost-cpp
```
Please look at these warnings from the conda-build link inspector: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=491541&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=1815

Are you using the header only boost libraries or statically linking? Or sometimes boost is only used for tests? Switch from static to dynamic linking or ensure that the run requirements excludes boost.

traversaro replied:
Only the header-only part of boost-cpp is used, but it is used in the public headers of `fmi4cpp` so if we do not include it, for downstream C++ users `find_package(fmi4cpp REQUIRED)` will fail. If you prefer, I can remove it, but as this is a pure C++ library I tought it was desirable to simplify as much as possible the usage of the library from the point of view of C++ users.

carterbox replied:
You should remove `boost-cpp` from the run requirements because it is not a run requirement, it is a build requirement. AFAIK, conda does not have a `build_exports` mechanism.

carterbox commented on recipes/fmi4cpp/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+
+build:
+  number: 0
+  # See https://github.com/conda-forge/staged-recipes/pull/18411#issuecomment-1073669556
+  skip: true  # [osx] 
```
How does getting a provisioned feedstock solve this error?

traversaro replied:
This is what we did in a similar occasion in https://github.com/conda-forge/staged-recipes/pull/13677/files#r562530228, I guess `staged-recipes` does not have the logic to handle enabling newer macOS SDKs. Feel free to suggest any alternative solution if you are aware of them, thanks!

carterbox replied:
OK.

carterbox commented on recipes/fmi4cpp/build.sh:
```diff
@@ -0,0 +1,13 @@
+
+cmake --build . --config Release
+cmake --build . --config Release --target install
+ctest --output-on-failure -C Release
```
Logs show that no tests are being run. Is that what you want? Existence tests are sufficient for conda-forge.

traversaro replied:
Thanks for noticing, whenever I can I also enable "regular" tests so that they can detect if there is something unexpected in the conda-forge-provided dependencies, but for sure it is not useful if not test runs.

traversaro replied:
Apparently fmi4cpp is using a custom CMake option to enable tests, I added it in https://github.com/conda-forge/staged-recipes/pull/18411/commits/29151a7f4d9aab0403e9b87629f16f41c85fc4fc and https://github.com/conda-forge/staged-recipes/pull/18411/commits/04781726527992129f0818672691b84598716b87 so that tests are actually running.

traversaro replied:
Apparently the 0.7.0 release of conan does not have tests. However, tests are present in the master version (see https://github.com/NTNU-IHB/FMI4cpp/blob/31cd95965da94652597e340c4b1b78c93fb830ec/CMakeLists.txt#L90) so if for you it is ok I would leave this here so that tests are going to run as soon as they are available.

carterbox replied:
Leaving this is fine since the intention is to have tests soon.

traversaro commented:
@conda-forge/help-c-cpp the comments have been addressed and the PR is again ready for review.

carterbox commented on recipes/fmi4cpp/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+  host:
+    - libzip
+    - boost-cpp
+  run:
+    - boost-cpp
```
```suggestion
```

traversaro commented:
Thanks @carterbox!
